### PR TITLE
Add 'requires' clauses for matrix template arguments to PETSc TS and SNES solvers.

### LINKED ISSUE #14840: Add C++20-style concepts to deal.II classes and functions
#14836 outlined a way towards annotating classes and functions with C++20-style `requires` clauses to check for concepts. Do this where appropriate.

### PR DESCRIPTION
**Pull request by bangerth**
This completes what I have in mind for #14840. 

@stefanozampini FYI.

Fixes #14840 .

### CODE REVIEW

File: include/deal.II/lac/petsc_snes.h [hunk 1]
```diff
@@ -233,13 +233,27 @@ namespace PETScWrappers
...
-                          std::constructible_from<VectorType, Vec>))
+  DEAL_II_CXX20_REQUIRES(
+    (concepts::is_dealii_petsc_vector_type<VectorType> ||
+     std::constructible_from<
+       VectorType,
+       Vec>)&&(concepts::is_dealii_petsc_matrix_type<PMatrixType> ||
+               std::constructible_from<
+                 PMatrixType,
+                 Mat>)&&(concepts::is_dealii_petsc_matrix_type<AMatrixType> ||
+                         std::constructible_from<AMatrixType, Mat>))
```
Comment (bangerth) [inline include/deal.II/lac/petsc_snes.h:256]: I'll readily admit that this is not wrapped in a way that makes it easy to read :-(

Comment (masterleinad) [reply include/deal.II/lac/petsc_snes.h:256]: You can always use `clang-format off`.

Comment (stefanozampini) [reply include/deal.II/lac/petsc_snes.h:256]: or update the clang-version :-)

**Review by stefanozampini — APPROVED** [review]

File: include/deal.II/lac/petsc_snes.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/lac/petsc_snes.h:256]: Not a battle for this PR to fight :-)

**Pull request merged**